### PR TITLE
Multiple JSON Pointers

### PR DESCRIPTION
**Pull request by fwolfsjaeger**
I also need the "unused" (not iterated) JSON, so I've extended your code to store this in a variable. I hope you don't mind.

### CODE REVIEW

**Comment by halaxa** [comment]
Thanks for being interested in this project. To be honest, I am concerned about 2 things.
1. Contents of `getUnusedJson` seem a little unintuitive. Doesn't it?
2. Filling up `storedJson` goes against the philosophy of this library. How do you prevent its size from getting out of available memory?

**Comment by fwolfsjaeger** [comment]
1. You're right, if we're iterating over an array the method will return something like "[,,,,,]". I will fix this so that it returns "[]".
2. It does consume additional memory, but other than writing to a temporary file I can't get that data without introducing a breaking change. I will check this again tomorrow.

**Comment by halaxa** [comment]
Let's talk essentials. What do you need the unused JSON parts for?

**Comment by fwolfsjaeger** [comment]
I have to split a large JSON containing meta-data, which I need once the data is processed.

Example:
```json
{
    "meta_data": {
        "total_rows": 5,
        "chunk_size": 2,
        "created_at": "2021-08-11T01:00:00+00:00"
    },
    "companies": [
        {
            "id": "1",
            "company": "ACME 1",
            "email": "office@acme1.com"
        },
        {
            "id": "2",
            "company": "ACME 2",
            "email": "office@acme2.com"
        },
        {
            "id": "3",
            "company": "ACME 3",
            "email": "office@acme3.com"
        },
        {
            "id": "4",
            "company": "ACME 4",
            "email": "office@acme4.com"
        },
        {
            "id": "5",
            "company": "ACME 5",
            "email": "office@acme5.com"
        }
    ]
}
```

Output 1:
```json
{
    "meta_data": {
        "total_rows": 5,
        "chunk_size": 3,
        "created_at": "2021-08-11T01:00:00+00:00",
        "total_files": 2,
        "file_number": 1
    },
    "companies": [
        {
            "id": "1",
            "company": "ACME 1",
            "email": "office@acme1.com"
        },
        {
            "id": "2",
            "company": "ACME 2",
            "email": "office@acme2.com"
        },
        {
            "id": "3",
            "company": "ACME 3",
            "email": "office@acme3.com"
        }
    ]
}
```

Output 2:
```json
{
    "meta_data": {
        "total_rows": 5,
        "chunk_size": 3,
        "created_at": "2021-08-11T01:00:00+00:00",
        "total_files": 2,
        "file_number": 2
    },
    "companies": [
        {
            "id": "4",
            "company": "ACME 4",
            "email": "office@acme4.com"
        },
        {
            "id": "5",
            "company": "ACME 5",
            "email": "office@acme5.com"
        }
    ]
}
```

**Comment by fwolfsjaeger** [comment]
What do you think about extending the parser to allow multiple JSON pointers? Parser.getIterator() would have to yield an array / object then.

**Comment by halaxa** [comment]
I have had exactly this idea for some time and I think that would solve your problem. In your case, you would specify `['/meta_data', '/companies']` instead of a single string with a JSON pointer. Is that correct?

`Parser::getIterator()` would not need to yield a multilevel structure though. I was wondering about method `Parser::getCurrentPath()` which would return matched path for the current element. It would simplify user iteration as well as implementation. What do you think?

**Comment by fwolfsjaeger** [comment]
It would solve my problem, right. Using Parser::getCurrentPath() sounds good as well. If you don't mind, I will amend the code to allow multiple pointers.

**Comment by fwolfsjaeger** [comment]
I have removed the "unused JSON logic" and extended the JSON pointer to allow multiple values. However, I have only added Parser::getMatchedJsonPointer(), which returns the original JSON pointer string.

Also, I have amended the Parser Unittest, which I could only test locally after upgrading to PHPunit 9.5.8. The PHPUnit upgrade is not part of this PR though.

What do you think?

**Comment by halaxa** [comment]
Can you please split it into 2 commits for easier orientation in the code? 1. Revert of the "unused" and 2. Your new implementation? Feel free to force push, thanks.

**Comment by fwolfsjaeger** [comment]
Sure, I've reverted my initial commit. Then I found an issue in the new implementation, which took quite a while to fix. I have added another Unittest for it.

**Comment by halaxa** [comment]
Thank you. I apologize for not replying. I have a kind of free-ish time of the year. I'll get to it as soon as I can. In the meantime, can you please amend README as well so it's in sync with these changes? It would be greatly appreciated. Thank you very much.

**Comment by fwolfsjaeger** [comment]
Sure, I've amended the README with a description of the multiple-pointer logic.

File: src/Parser.php [hunk 1]
```diff
@@ -33,39 +33,104 @@ class Parser implements \IteratorAggregate, PositionAware
...
+    private function buildJsonPointerPaths(array $jsonPointers)
+    {
+        $jsonPointers = array_values($jsonPointers);
+
+        foreach ($jsonPointers as $jsonPointerEl) {
+            if (preg_match('_^(/(([^/~])|(~[01]))*)*$_', $jsonPointerEl) === 0) {
+                throw new InvalidArgumentException(sprintf("Given value '%s' of \$jsonPointer is not valid JSON Pointer", $jsonPointerEl));
+            }
+
+            $intersectingJsonPointers = array_filter($jsonPointers, static function($el) use ($jsonPointerEl) {
```
Comment (halaxa) [inline src/Parser.php:73]: What if the pointers with wildcard intersect? Let's say `['/list/-/items', '/list/1/items', '/list/1/items/item']`?

File: src/Parser.php [hunk 2]
```diff
@@ -93,38 +158,44 @@ public function getIterator()
...
 
         // local variables for faster name lookups
         $lexer = $this->lexer;
-        $jsonPointerPath = $this->jsonPointerPath;
-        
+
         foreach ($lexer as $token) {
+            if ($currentPathChanged) {
+                $currentPathChanged = false;
+                $jsonPointerPath = $this->getMatchingJsonPointerPath($currentPath);
```
Comment (halaxa) [inline src/Parser.php:183]: I strive to call as few functions as possible in this loop. Also, the function seems overly complex to do the job. To compare current path against multiple JSON pointers I was thinking something along the lines of `isset($jsonPointers[$currentPath]) || isset($jsonPointers[preg_replace('~/[0-9]+[/$]~', '-', $currentPath)])` (treat is as pseudocode to put an idea across).

File: src/Parser.php (hunk 1 above)
Comment (fwolfsjaeger) [reply src/Parser.php:73]: I've amended the code to also check for wildcards

File: src/Parser.php (hunk 2 above)
Comment (fwolfsjaeger) [reply src/Parser.php:204]: The function is that complex because we also need the closest match in case there is no full match yet, otherwise it would discard a JSON pointer path too soon. I've tried a simpler approach using RegExp, but so far I could not find any other functional solution.

**Comment by fwolfsjaeger** [comment]
Conflicts resolved by rebasing your changes ...

**Comment by halaxa** [comment]
Thanks. I'll try to get to it asap.

**Comment by fwolfsjaeger** [comment]
> Yes, along with a solution. I have to keep track of an additional $currentPathWildcard variable though.

Yes, along with a solution. I have to keep track of an additional $currentPathWildcard variable though.

**Comment by fwolfsjaeger** [comment]
> Can you commit the failing test?

Yes, along with a solution. I have to keep track of an additional $currentPathWildcard variable though.

**Comment by fwolfsjaeger** [comment]
@halaxa I've merged your latest changes and resolved the conflicts.

**Comment by halaxa** [comment]
Thanks. Does it mean it is ready to merge from your side?

**Comment by fwolfsjaeger** [comment]
It works and all tests pass, but I guess it would make sense to run some performance-tests. Do you have existing tests for that?

**Comment by halaxa** [comment]
Good point. Just run `make performance-tests` (uses docker in the background) or `composer performance-tests` (no docker). You can compare the results with and without your feature. I've done some performance tweaking on Lexer so that `0.8.0` will have about 150% performance of `0.7.*` (what took 10s will now take about 7s). So slight performance decrease of this feature will hopefully fit in so that it stays on the plus side.

**Comment by fwolfsjaeger** [comment]
Thanks for the hint! I've run `composer performance-tests` 10 - 20 times for each branch and didn't notice a significant difference.

**Comment by halaxa** [comment]
Thanks. Can you, please, look at the CHANGELOG and accommodate the changes - which are coming with version 0.8.0 -accordingly?

**Comment by fwolfsjaeger** [comment]
My branch contains all changes from 0.8.0 now.

**Comment by halaxa** [comment]
You are a very zealous coder. So before you commit anything, let's talk first. :) What if we added option `pointers` only accepting arrays (in case of multiple pointers) and `pointer` would keep only accepting strings to enhance clarity and readability to end-users?

**Comment by fwolfsjaeger** [comment]
We would have to add another array key to the "from" methods as well. Also, for the Parser class it doesn't make any difference.

What about adding a copy of the Items class - named something like DiverseItems or VaryingItems - instead of just a dedicated constructor argument?
- the method getMatchedJsonPointer() is only required there
- we could use proper type hints

**Comment by halaxa** [comment]
> We would have to add another array key to the "from" methods as well.

Yes, they are all checked in one place.

> Also, for the `Parser` class, it doesn't make any difference.

Also yes - no problem. This would be just convenient naming in `Items` options for the developers. I'd keep `Parser`'s constructor as it is.

**Comment by fwolfsjaeger** [comment]
I am fetching data from a REST endpoint that dynamically creates the data based on a last-modified timestamp. Unfortunately, in that case I don't know the order. However, I've tried to change it to check the performance, but I could not get it to work like that for multiple JSON pointers.

I could remove the check for intersecting JSON pointers, but this really doesn't cost much time/resources. However, I can add a check so that the intersection check is only executed if multiple pointers are provided.

Also, we only need to call the method getMatchingJsonPointerPath() if there are multiple JSON pointers. For huge JSON documents with only a single pointer this might actually make a difference. With those changes the performance should be the same as with your original code, unless there are multiple JSON pointers.

**Comment by halaxa** [comment]
I was only reviving the idea to make the parser less complex. I wasn't concerned about the performance. But since your own use case would be the 0.01 %, let's go with it. Let me just do some final reviews ;)

**Comment by fwolfsjaeger** [comment]
Some additional small changes:
1. I've renamed the variables I've introduced for consistency.
2. I've reverted a backwards-incompatible change I've made: JsonMachine\Parser::getJsonPointer() returns a string just like before again.
3. I've introduced the check variable $singleJsonPointer so that the performance for single pointer operations is not affected by my changes.

**Comment by halaxa** [comment]
`Items::getMatchedJsonPointer()` should stay, shouldn't it?

**Comment by halaxa** [comment]
See [#issuecomment-896821976](#issuecomment-896821976)

**Comment by fwolfsjaeger** [comment]
Items::getJsonPointer() calls Parser::getJsonPointer(), which returns the current JSON pointer.

**Comment by halaxa** [comment]
Tomorrow :)

**Comment by halaxa** [comment]
> Items::getJsonPointer() calls Parser::getJsonPointer(), which returns the current JSON pointer.

Yes, however, that's a major change in behavior. Previously it was just a simple property getter. That's why I think it should be:

**`getJsonPointer(): string`** - keep the behavior (deprecate the method?)

**`getJsonPointers(): array`** - returns an array of all provided JSON pointers (simple property getter)

**`getCurrentJsonPointer(): string`** - returns the currently iterated path in the document in JSON pointer format (null or Exception outside of a `foreach`?)

**`getMatchedJsonPointer(): string`** - returns currently matched JSON pointer - one of the provided ones (null or Exception outside of a `foreach`?)

What do you think? Do better names come to mind?

**Comment by fwolfsjaeger** [comment]
Actually, I've broken backwards compatibility by returning an array of all JSON pointers in Parser::getJsonPointer(). With my latest changes Parser::getJsonPointer() returns the current JSON pointer as a string, just like before.

Current state:
`Parser::getJsonPointer(): string` - returns the current JSON pointer (if there is only one it behaves just like before)
`Parser::getJsonPointers(): array` - returns an array of all JSON pointers
`Parser::getJsonPointerPath(): string` - returns the current JSON pointer path (if there is only one it behaves just like before)
`Parser::getJsonPointerPaths(): array` - returns an array of all JSON pointers paths

To make it clear that the current JSON pointer [path] is returned we could deprecate the existing methods and add them with other names. I would keep the current state, but it is your project and therefor your call ;)

**Comment by halaxa** [comment]
`Parser::getJsonPointerPath(): string` was there only for testing reasons. It was not meant to provide any useful info as a part of the API. I think it just remained there as an artifact of initial TDD. It uncovers internal implementation and I guess no one needs this format for anything. Therefore I would not introduce `Parser::getJsonPointerPaths(): array` and deprecate `Parser::getJsonPointerPath(): string`.

`Parser::getCurrentJsonPointer(): string` and `Parser::getMatchedJsonPointer(): string` are a different story. They specifically say what exactly they do and are very useful in the `foreach` when someone iterates over multiple JSON Pointers or uses `-` as a wildcard. They tell you the current position and which JSON Pointer exactly matched this position. They also don't introduce a new format of JSON document path but rather provide already known JSON Pointer. That can be directly put back to the `pointer` option of any `from*` factories without any modification if one wishes.

Does it make sense like this?

**Comment by fwolfsjaeger** [comment]
Makes sense, I'll make the following changes then:
1. deprecate Parser::getJsonPointer()
2. remove Parser::getJsonPointerPath()
3. add Parser::getCurrentJsonPointer() returning the current path (no wildcards) in JSON pointer format
4. and Parser::getMatchedJsonPointer() returning the currently matched JSON pointer

**Comment by halaxa** [comment]
Awsome, thanks. Also, I think that `Parser::getCurrentJsonPointer()` and `Parser::getMatchedJsonPointer()` should throw if used outside of a loop.

**Comment by fwolfsjaeger** [comment]
I agree, both throw a JsonMachineException if called outside the foreach loop.

I've added 3 new tests, one to check the return values of the 2 new methods and two more to check for the exceptions. They have to be separate as checking the exeption doesn't work otherwise.

**Comment by halaxa** [comment]
Thanks. I don't see `Parser::getJsonPointers(): array` though. Did you push everything?

**Comment by fwolfsjaeger** [comment]
I have not added `Parser::getJsonPointers()` as it would just return the same array that has been provided. If you want, I can add it though.

**Comment by halaxa** [comment]
Yes, that's the [goal](https://github.com/halaxa/json-machine/pull/61#issuecomment-1013354417). The users should be able to get what they put in. Thanks

**Comment by halaxa** [comment]
I added dome edge case tests to the new getters and one of them is failing. Can you please extend the functionality so it conforms to the JSON Pointer standard? [Root level iteration is actually an empty-string JSON Pointer](https://datatracker.ietf.org/doc/html/rfc6901#section-5). Thanks.

**Comment by fwolfsjaeger** [comment]
You are right, I've amended the method.

**Comment by halaxa** [comment]
Not sure if you were notified. Can you check [this comment](https://github.com/halaxa/json-machine/pull/61/commits/80779cedc06e671b77cd926a4278e3817c682126#diff-0415811a39e7e77b558673e5d670e72c3d5ac548ec103cd992c97bb4d7f00f7dL327), please?

**Comment by fwolfsjaeger** [comment]
Sorry, but I got no notification and can't see any comment.

**Comment by halaxa** [comment]
No problem, I was just asking why did you remove `['/', '{"":{"c":1,"d":2}}', ['/', '/']]` test case in `ParserTest`?

**Comment by halaxa** [comment]
Awesome, thank you. I'll try to do the final review asap.

**Comment by halaxa** [comment]
Ok, thanks.

**Comment by halaxa** [comment]
I have a question. I am probably missing something. What is `Parser::$jsonPointer` used for at this moment? Doesn't its purpose overlap with  `Parser::$currentJsonPath`?

**Comment by fwolfsjaeger** [comment]
`Parser::$jsonPointer` is the matched JSON pointer, it can be fetched using `Parser::getMatchedJsonPointer()`: eg. /items/-
`Parser::$currentJsonPath` is the current JSON path, it's needed for `Parser::getCurrentJsonPointer()`: eg. /items/0

**Comment by halaxa** [comment]
OK, so then
- `Parser::getJsonPointer()` completely changed its meaning. It should preserve its original function.
- `Parser::$jsonPointer` should be renamed to something like `Parser::$matchedJsonPointer` to better reflect its purpose.

Can you do it or should I?

**Comment by fwolfsjaeger** [comment]
`Parser::getJsonPointer()` still returns the passed JSON pointer when inside the iteration. But fine, I have changed it so that it always just returns the first JSON pointer, it won't work with multiple pointers anymore.
`Parser::$jsonPointer` is named `Parser::$matchedJsonPointer` now.

**Comment by halaxa** [comment]
> it won't work with multiple pointers anymore.

That's ok as it should keep previous functionality. I plan to make `0.9.0` before `1.0.0` so I'd like to keep BC.

> `Parser::$jsonPointer` is named `Parser::$matchedJsonPointer` now.

Thanks :)
